### PR TITLE
Added no_alloc variant

### PR DESCRIPTION
**Pull request by windkit**
### Description

JErasure Library uses lots of heap allocation for temporary structure.
The large number of small allocation/deallocation may have performance impact.
When used in NIF, some memory issue is encountered (Random Segmentation Fault)
### Purposed Solution

Added a separate code path which avoid memory allocation (structures are allocated by caller beforehand), putting temporary structure in stack.
### Related Issue

https://github.com/leo-project/leofs/issues/440
### Relate PR

https://github.com/leo-project/jerasure/pull/3
https://github.com/leo-project/jerasure/pull/4
https://github.com/leo-project/leo_erasure/pull/1

### CODE REVIEW

**Comment by mocchira** [comment]
@windkit 
At first,
Keep original indents if you want to PR also to the upstream.
Complying with a original coding rule is one of the most important thing.

I'll check the remain part after rolling back to the original indent.

**Comment by mocchira** [comment]
@windkit there are still some modifications for fixing indents.
Yea I know you're right ( the original indent are broken at some places ).
but to make this PR simpler, 
I'd recommend you dividing modifications for fixing broken indents into another PR.

File: Examples/encoder.c [hunk 1]
```diff
@@ -583,18 +621,17 @@ int main (int argc, char **argv) {
 		fclose(fp2);
 	}
 
+	/* Calculate rate in MB/sec and print */
+	timing_set(&t2);
+	tsec = timing_delta(&t1, &t2);
+	printf("Encoding (MB/sec): %0.10f\n", (((double) size)/1024.0/1024.0)/totalsec);
+	printf("En_Total (MB/sec): %0.10f\n", (((double) size)/1024.0/1024.0)/tsec);
```
Comment (mocchira) [inline Examples/encoder.c:624]: Why did you change the place where the code block for calculating rate was ( originally placing after freeing allocated memory ) ?

**Comment by windkit** [comment]
@mocchira I have rebased the PR, please check.

**Comment by mocchira** [comment]
@windkit 
Naming helper functions for no alloc version in `jerasure.c` like `jerasure_matrix_decode_data` the original name  is `jerasure_matrix_decode` and `jerasure_schedule_decode_data_lazy` the original name is `jerasure_schedule_decode_lazy` are difficult to understand to me.
How about simply suffixing by `_noalloc` as well as others?

**Comment by windkit** [comment]
@mocchira `_data` means it would only decode missing data blocks (ignore any missing code blocks)
While the original one would decode all missing data+code blocks.

**Comment by mocchira** [comment]
@windkit which means 
Can that part of code be divided into another PR? ( not related to this PR's purpose using stack instead of heap )
if so, please get rid of that part to make this PR more simpler.

**Comment by windkit** [comment]
@mocchira True, I will split it into two

**Comment by windkit** [comment]
@mocchira I have separated it and created a PR at https://github.com/leo-project/jerasure/pull/4

It depends on this PR and leo-project/leo_erasure#1 depends on both PRs

**Comment by mocchira** [comment]
@windkit 
why are there something overlapped code?
( I'd expect #3 and #4 will have no duplicated code each other. )

To avoid confusion,
#4 should be closed while #3 is in-progress.

**Comment by windkit** [comment]
@mocchira Selective decoding make use of the no_alloc methods, that is the reason.
Or could you suggest how should I create the PRs?

**Comment by mocchira** [comment]
@windkit 
It seems difficult that #4 include commits #3 have done.
As I said at the previous comment,
Please close #4 while #3 is in-progress.

**Comment by windkit** [comment]
@mocchira Oh? I thought it would "identify" duplicate commits after #3 is merged

**Comment by mocchira** [comment]
LGTM.
